File: frontend/src/api/sources.ts
import { ApiError } from "./auth";

const BASE = import.meta.env.VITE_API_BASE_URL || "http://localhost:8000";

export interface SourceRead {
  id: string;
  owner_id: string;
  title: string;
  archive_reference: string | null;
  url: string | null;
  reliability_comment: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface SourceCreate {
  title: string;
  archive_reference?: string | null;
  url?: string | null;
  reliability_comment?: string | null;
  notes?: string | null;
}

export interface SourceUpdate {
  title?: string;
  archive_reference?: string | null;
  url?: string | null;
  reliability_comment?: string | null;
  notes?: string | null;
}

export interface PersonSourceRead {
  id: string;
  person_id: string;
  source_id: string;
  comment: string | null;
  created_at: string;
  updated_at: string;
}

export interface RelationshipSourceRead {
  id: string;
  relationship_id: string;
  source_id: string;
  comment: string | null;
  created_at: string;
  updated_at: string;
}

async function handleResponse<T>(res: Response): Promise<T> {
  if (res.ok) {
    return res.json() as Promise<T>;
  }
  let detail = `HTTP ${res.status}`;
  try {
    const body = (await res.json()) as { detail?: string };
    if (typeof body.detail === "string") detail = body.detail;
  } catch {
    // ignore parse errors
  }
  throw new ApiError(res.status, detail);
}

function jsonHeaders(token: string): HeadersInit {
  return {
    "Content-Type": "application/json",
    Authorization: `Bearer ${token}`,
  };
}

async function handleNoContent(res: Response): Promise<void> {
  if (res.ok) return;
  let detail = `HTTP ${res.status}`;
  try {
    const body = (await res.json()) as { detail?: string };
    if (typeof body.detail === "string") detail = body.detail;
  } catch {
    // ignore
  }
  throw new ApiError(res.status, detail);
}

export async function listSources(token: string): Promise<SourceRead[]> {
  let res: Response;
  try {
    res = await fetch(`${BASE}/api/v1/sources`, {
      headers: { Authorization: `Bearer ${token}` },
    });
  } catch {
    throw new ApiError(0, "Backend недоступен");
  }
  return handleResponse<SourceRead[]>(res);
}

export async function listPersonSourceLinks(
  token: string,
): Promise<PersonSourceRead[]> {
  let res: Response;
  try {
    res = await fetch(`${BASE}/api/v1/sources/person-links`, {
      headers: { Authorization: `Bearer ${token}` },
    });
  } catch {
    throw new ApiError(0, "Backend недоступен");
  }
  return handleResponse<PersonSourceRead[]>(res);
}

export async function listRelationshipSourceLinks(
  token: string,
): Promise<RelationshipSourceRead[]> {
  let res: Response;
  try {
    res = await fetch(`${BASE}/api/v1/sources/relationship-links`, {
      headers: { Authorization: `Bearer ${token}` },
    });
  } catch {
    throw new ApiError(0, "Backend недоступен");
  }
  return handleResponse<RelationshipSourceRead[]>(res);
}

export async function createSource(
  token: string,
  payload: SourceCreate,
): Promise<SourceRead> {
  let res: Response;
  try {
    res = await fetch(`${BASE}/api/v1/sources`, {
      method: "POST",
      headers: jsonHeaders(token),
      body: JSON.stringify(payload),
    });
  } catch {
    throw new ApiError(0, "Backend недоступен");
  }
  return handleResponse<SourceRead>(res);
}

export async function updateSource(
  token: string,
  sourceId: string,
  payload: SourceUpdate,
): Promise<SourceRead> {
  let res: Response;
  try {
    res = await fetch(`${BASE}/api/v1/sources/${sourceId}`, {
      method: "PATCH",
      headers: jsonHeaders(token),
      body: JSON.stringify(payload),
    });
  } catch {
    throw new ApiError(0, "Backend недоступен");
  }
  return handleResponse<SourceRead>(res);
}

export async function deleteSource(
  token: string,
  sourceId: string,
): Promise<void> {
  let res: Response;
  try {
    res = await fetch(`${BASE}/api/v1/sources/${sourceId}`, {
      method: "DELETE",
      headers: { Authorization: `Bearer ${token}` },
    });
  } catch {
    throw new ApiError(0, "Backend недоступен");
  }
  return handleNoContent(res);
}

export async function linkSourceToPerson(
  token: string,
  sourceId: string,
  personId: string,
): Promise<PersonSourceRead> {
  let res: Response;
  try {
    res = await fetch(
      `${BASE}/api/v1/sources/${sourceId}/persons/${personId}`,
      {
        method: "POST",
        headers: { Authorization: `Bearer ${token}` },
      },
    );
  } catch {
    throw new ApiError(0, "Backend недоступен");
  }
  return handleResponse<PersonSourceRead>(res);
}

export async function unlinkSourceFromPerson(
  token: string,
  sourceId: string,
  personId: string,
): Promise<void> {
  let res: Response;
  try {
    res = await fetch(
      `${BASE}/api/v1/sources/${sourceId}/persons/${personId}`,
      {
        method: "DELETE",
        headers: { Authorization: `Bearer ${token}` },
      },
    );
  } catch {
    throw new ApiError(0, "Backend недоступен");
  }
  return handleNoContent(res);
}

export async function linkSourceToRelationship(
  token: string,
  sourceId: string,
  relationshipId: string,
): Promise<RelationshipSourceRead> {
  let res: Response;
  try {
    res = await fetch(
      `${BASE}/api/v1/sources/${sourceId}/relationships/${relationshipId}`,
      {
        method: "POST",
        headers: { Authorization: `Bearer ${token}` },
      },
    );
  } catch {
    throw new ApiError(0, "Backend недоступен");
  }
  return handleResponse<RelationshipSourceRead>(res);
}

export async function unlinkSourceFromRelationship(
  token: string,
  sourceId: string,
  relationshipId: string,
): Promise<void> {
  let res: Response;
  try {
    res = await fetch(
      `${BASE}/api/v1/sources/${sourceId}/relationships/${relationshipId}`,
      {
        method: "DELETE",
        headers: { Authorization: `Bearer ${token}` },
      },
    );
  } catch {
    throw new ApiError(0, "Backend недоступен");
  }
  return handleNoContent(res);
}

File: backend/app/api/routes/sources.py
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Response, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.security import get_current_user
from app.db.session import get_db
from app.models.person import Person
from app.models.relationship import Relationship
from app.models.source import PersonSource, RelationshipSource, Source
from app.models.user import User
from app.schemas.source import (
    PersonSourceRead,
    RelationshipSourceRead,
    SourceCreate,
    SourceRead,
    SourceUpdate,
)

router = APIRouter()
DbSession = Annotated[Session, Depends(get_db)]
CurrentUser = Annotated[User, Depends(get_current_user)]


def get_owned_source(source_id: uuid.UUID, owner_id: uuid.UUID, db: Session) -> Source:
    source = db.scalar(
        select(Source).where(
            Source.id == source_id,
            Source.owner_id == owner_id,
        )
    )
    if source is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Источник не найден")
    return source


def get_owned_person(person_id: uuid.UUID, owner_id: uuid.UUID, db: Session) -> Person:
    person = db.scalar(
        select(Person).where(
            Person.id == person_id,
            Person.owner_id == owner_id,
        )
    )
    if person is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Персона не найдена")
    return person


def get_owned_relationship(
    relationship_id: uuid.UUID,
    owner_id: uuid.UUID,
    db: Session,
) -> Relationship:
    relationship = db.scalar(
        select(Relationship).where(
            Relationship.id == relationship_id,
            Relationship.owner_id == owner_id,
        )
    )
    if relationship is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Связь не найдена",
        )
    return relationship


def get_person_source_link(
    source_id: uuid.UUID,
    person_id: uuid.UUID,
    db: Session,
) -> PersonSource | None:
    return db.scalar(
        select(PersonSource).where(
            PersonSource.source_id == source_id,
            PersonSource.person_id == person_id,
        )
    )


def get_relationship_source_link(
    source_id: uuid.UUID,
    relationship_id: uuid.UUID,
    db: Session,
) -> RelationshipSource | None:
    return db.scalar(
        select(RelationshipSource).where(
            RelationshipSource.source_id == source_id,
            RelationshipSource.relationship_id == relationship_id,
        )
    )


@router.post("", response_model=SourceRead, status_code=status.HTTP_201_CREATED)
def create_source(payload: SourceCreate, db: DbSession, current_user: CurrentUser) -> Source:
    source = Source(owner_id=current_user.id, **payload.model_dump())
    db.add(source)
    db.commit()
    db.refresh(source)
    return source


@router.get("", response_model=list[SourceRead])
def list_sources(db: DbSession, current_user: CurrentUser) -> list[Source]:
    return list(
        db.scalars(
            select(Source)
            .where(Source.owner_id == current_user.id)
            .order_by(Source.title, Source.id)
        )
    )


@router.get("/person-links", response_model=list[PersonSourceRead])
def list_person_source_links(
    db: DbSession,
    current_user: CurrentUser,
) -> list[PersonSource]:
    return list(
        db.scalars(
            select(PersonSource)
            .join(Source, PersonSource.source_id == Source.id)
            .join(Person, PersonSource.person_id == Person.id)
            .where(
                Source.owner_id == current_user.id,
                Person.owner_id == current_user.id,
            )
            .order_by(PersonSource.created_at.desc())
        )
    )


@router.get("/relationship-links", response_model=list[RelationshipSourceRead])
def list_relationship_source_links(
    db: DbSession,
    current_user: CurrentUser,
) -> list[RelationshipSource]:
    return list(
        db.scalars(
            select(RelationshipSource)
            .join(Source, RelationshipSource.source_id == Source.id)
            .join(Relationship, RelationshipSource.relationship_id == Relationship.id)
            .where(
                Source.owner_id == current_user.id,
                Relationship.owner_id == current_user.id,
            )
            .order_by(RelationshipSource.created_at.desc())
        )
    )


@router.get("/{source_id}", response_model=SourceRead)
def read_source(source_id: uuid.UUID, db: DbSession, current_user: CurrentUser) -> Source:
    return get_owned_source(source_id=source_id, owner_id=current_user.id, db=db)


@router.patch("/{source_id}", response_model=SourceRead)
def update_source(
    source_id: uuid.UUID,
    payload: SourceUpdate,
    db: DbSession,
    current_user: CurrentUser,
) -> Source:
    source = get_owned_source(source_id=source_id, owner_id=current_user.id, db=db)
    for field, value in payload.model_dump(exclude_unset=True).items():
        setattr(source, field, value)

    db.commit()
    db.refresh(source)
    return source


@router.delete("/{source_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_source(source_id: uuid.UUID, db: DbSession, current_user: CurrentUser) -> Response:
    source = get_owned_source(source_id=source_id, owner_id=current_user.id, db=db)
    db.delete(source)
    db.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post(
    "/{source_id}/persons/{person_id}",
    response_model=PersonSourceRead,
    status_code=status.HTTP_201_CREATED,
)
def link_source_to_person(
    source_id: uuid.UUID,
    person_id: uuid.UUID,
    db: DbSession,
    current_user: CurrentUser,
) -> PersonSource:
    get_owned_source(source_id=source_id, owner_id=current_user.id, db=db)
    get_owned_person(person_id=person_id, owner_id=current_user.id, db=db)

    if get_person_source_link(source_id=source_id, person_id=person_id, db=db) is not None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Источник уже привязан к этой персоне",
        )

    link = PersonSource(source_id=source_id, person_id=person_id)
    db.add(link)
    db.commit()
    db.refresh(link)
    return link


@router.delete(
    "/{source_id}/persons/{person_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def unlink_source_from_person(
    source_id: uuid.UUID,
    person_id: uuid.UUID,
    db: DbSession,
    current_user: CurrentUser,
) -> Response:
    get_owned_source(source_id=source_id, owner_id=current_user.id, db=db)
    get_owned_person(person_id=person_id, owner_id=current_user.id, db=db)

    link = get_person_source_link(source_id=source_id, person_id=person_id, db=db)
    if link is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Привязка источника к персоне не найдена",
        )

    db.delete(link)
    db.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post(
    "/{source_id}/relationships/{relationship_id}",
    response_model=RelationshipSourceRead,
    status_code=status.HTTP_201_CREATED,
)
def link_source_to_relationship(
    source_id: uuid.UUID,
    relationship_id: uuid.UUID,
    db: DbSession,
    current_user: CurrentUser,
) -> RelationshipSource:
    get_owned_source(source_id=source_id, owner_id=current_user.id, db=db)
    get_owned_relationship(relationship_id=relationship_id, owner_id=current_user.id, db=db)

    if (
        get_relationship_source_link(
            source_id=source_id,
            relationship_id=relationship_id,
            db=db,
        )
        is not None
    ):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Источник уже привязан к этой связи",
        )

    link = RelationshipSource(source_id=source_id, relationship_id=relationship_id)
    db.add(link)
    db.commit()
    db.refresh(link)
    return link


@router.delete(
    "/{source_id}/relationships/{relationship_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def unlink_source_from_relationship(
    source_id: uuid.UUID,
    relationship_id: uuid.UUID,
    db: DbSession,
    current_user: CurrentUser,
) -> Response:
    get_owned_source(source_id=source_id, owner_id=current_user.id, db=db)
    get_owned_relationship(relationship_id=relationship_id, owner_id=current_user.id, db=db)

    link = get_relationship_source_link(
        source_id=source_id,
        relationship_id=relationship_id,
        db=db,
    )
    if link is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Привязка источника к связи не найдена",
        )

    db.delete(link)
    db.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)
